### PR TITLE
8286625: C2 fails with assert(!n->is_Store() && !n->is_LoadStore()) failed: no node with a side effect

### PR DESCRIPTION
**Pull request by rwestrel**
It's another case where because of overunrolling, the main loop is
never executed but not optimized out and the type of some
CastII/ConvI2L for a range check conflicts with the type of its input
resulting in a broken graph for the main loop.

This is supposed to have been solved by skeleton predicates. There's
indeed a predicate that should catch that the loop is unreachable but
it doesn't constant fold. The shape of the predicate is:

(CmpUL (SubL 15 (ConvI2L (AddI (CastII int:>=1) 15) minint..maxint)) 16)

I propose adding a CastII, that is in this case:

(CmpUL (SubL 15 (ConvI2L (CastII (AddI (CastII int:>=1) 15) 0..max-1) minint..maxint)) 16)

The justification for the CastII is that the skeleton predicate is a
predicate for a specific iteration of the loop. That iteration of the
loop must be in the range of the iv Phi.

With the extra CastII, the AddI can be pushed through the CastII and
ConvI2L and the check constant folds. Actually, with the extra CastII,
the predicate is not implemented with a CmpUL but a CmpU because the
code can tell there's no risk of overflow (I did force the use of
CmpUL as an experiment and the CmpUL does constant fold)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8286625](https://bugs.openjdk.org/browse/JDK-8286625): C2 fails with assert(!n->is_Store() && !n->is_LoadStore()) failed: no node with a side effect

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [8f4a999a](https://git.openjdk.java.net/jdk/pull/8996/files/8f4a999a97a41fef360bc14b3ea3a2f9a92aa237)
 * [Christian Hagedorn](https://openjdk.java.net/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [8f4a999a](https://git.openjdk.java.net/jdk/pull/8996/files/8f4a999a97a41fef360bc14b3ea3a2f9a92aa237)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8996/head:pull/8996` \
`$ git checkout pull/8996`

Update a local copy of the PR: \
`$ git checkout pull/8996` \
`$ git pull https://git.openjdk.java.net/jdk pull/8996/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8996`

View PR using the GUI difftool: \
`$ git pr show -t 8996`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8996.diff">https://git.openjdk.java.net/jdk/pull/8996.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/compiler/loopopts/TestOverUnrolling2.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8286625
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/loopopts/TestOverUnrolling2.java:26]:
```suggestion
 * @bug 8286625
 * @key stress
```

**Review by TobiHartmann — APPROVED** [review]
Looks reasonable to me. I submittest testing and will report back once it passed.

**Review by chhagedorn — APPROVED** [review]
Looks good!

**Comment by TobiHartmann** [comment]
All tests passed.

**Comment by rwestrel** [comment]
@TobiHartmann @chhagedorn thanks for the reviews

**Pull request closed without merging**